functions/endChat.ts

```diff
@@ -63,6 +63,105 @@ const getEndChatMessage = (event: Body): string => {
   return 'User left the conversation';
 };
 
+/**
+ * End a task by updating their assignment status.
+ *
+ * It also sends a message indicating that the user has left the conversation
+ * if appropriate.
+ *
+ * @returns channelCleanupRequired
+ */
+const updateTaskAssignmentStatus = async (
+  taskSid: string,
+  channelSid: string,
+  context: Context<EnvVars>,
+  event: Body,
+) => {
+  const client = context.getTwilioClient();
+  let channelCleanupRequired = false;
+
+  // Fetch the Task to 'cancel' or 'wrapup'
+  const task = await client.taskrouter
+    .workspaces(context.TWILIO_WORKSPACE_SID)
+    .tasks(taskSid)
+    .fetch();
+
+  // Send a Message indicating user left the conversation
+  if (task.assignmentStatus === 'assigned') {
+    const endChatMessage = getEndChatMessage(event);
+    await context
+      .getTwilioClient()
+      .chat.services(context.CHAT_SERVICE_SID)
+      .channels(channelSid)
+      .messages.create({
+        body: endChatMessage,
+        from: 'Bot',
+        xTwilioWebhookEnabled: 'true',
+      });
+  }
+
+  // Update the task assignmentStatus
+  const updateAssignmentStatus = (assignmentStatus: TaskInstance['assignmentStatus']) =>
+    client.taskrouter
+      .workspaces(context.TWILIO_WORKSPACE_SID)
+      .tasks(taskSid)
+      .update({ assignmentStatus });
+
+  switch (task.assignmentStatus) {
+    case 'reserved':
+    case 'pending': {
+      await updateAssignmentStatus('canceled');
+      channelCleanupRequired = true;
+      break;
+    }
+    case 'assigned': {
+      await updateAssignmentStatus('wrapping');
+      break;
+    }
+    default:
+  }
+
+  return channelCleanupRequired;
+};
+
+/**
+ * End contact task or post-survey task associated to the given channel.
+ *
+ * @returns channelCleanupRequired
+ */
+const endContactOrPostSurvey = async (
+  channelAttributes: any,
+  context: Context<EnvVars>,
+  event: Body,
+) => {
+  const { taskSid, surveyTaskSid } = channelAttributes;
+  const { channelSid } = event;
+  const updateTaskPromises: Promise<boolean>[] = [];
+
+  if (taskSid) {
+    const updateContactTask = updateTaskAssignmentStatus(
+      taskSid,
+      channelSid as string,
+      context,
+      event,
+    );
+    updateTaskPromises.push(updateContactTask);
+  }
+
+  if (surveyTaskSid) {
+    const updatePostSurveyTask = updateTaskAssignmentStatus(
+      surveyTaskSid,
+      channelSid as string,
+      context,
+      event,
+    );
+    updateTaskPromises.push(updatePostSurveyTask);
+  }
+
+  const resolvedPromises = await Promise.all(updateTaskPromises);
+  return resolvedPromises.some((channelCleanupRequired: boolean) => channelCleanupRequired);
```

Comment:
I think this can cause race conditions if the original task does not exists any more. I don't think this will make the function to entirely fail, but just in case can we use `Promise.allSettled()` instead?